add tf2 broadcaster and listener tutorials (#1715)
ros2/ros2_documentation: source/Tutorials/Tf2/Writing-A-Tf2-Broadcaster-Py.rst

.. _WritingATf2BroadcasterPy:

Writing a tf2 broadcaster (Python)
==================================

**Goal:** Learn how to broadcast the state of a robot to tf2.

**Tutorial level:** Intermediate

**Time:** 15 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

In the next two tutorials we will write the code to reproduce the demo from the :ref:`Introduction to tf2 <IntroToTf2>` tutorial.
After that, the following tutorials focus on extending the demo with more advanced tf2 features.

Prerequisites
-------------

In previous tutorials, you learned how to :ref:`create a workspace <ROS2Workspace>` and :ref:`create a package <CreatePkg>`.
You also have created the ``learning_tf2_py`` package where we will continue working on.

Tasks
-----

1 Write the broadcaster node
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Let's first create the source files.
Go to the ``learning_tf2_py`` package we created in the previous tutorial.
Inside the ``src/learning_tf2_py/learning_tf2_py`` directory, create a new file called
``turtle_tf2_broadcaster.py`` and paste the following code within:

> .. code-block:: python

Reviewer: Is there a way to link to the Github code? What if the code changes?

Author: I don't know if there is a direct way to embed code directly from the Github. Do think simply linking the file will work?

Author: @ahcorde the code in other tutorials is also present on the tutorial pages themselves as in https://github.com/ros2/ros2_documentation/blob/rolling/source/Tutorials/Writing-A-Simple-Py-Publisher-And-Subscriber.rst#2-write-the-publisher-node. However, I can add additional line on how to download them directly from github.

                    {'turtlename': 'turtle1'}
                ]
            ),
        ])

2.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

First we import required modules from the ``launch`` and ``launch_ros`` packages.

.. code-block:: python

    from launch import LaunchDescription
    from launch_ros.actions import Node

Now we run our nodes that start the turtlesim simulation and broadcast ``turtle1`` state to the tf2 using our ``turtle_tf2_broadcaster`` node.

.. code-block:: python

    Node(
        package='turtlesim',
        executable='turtlesim_node',
        name='sim'
    ),
    Node(
        package='learning_tf2_py',
        executable='turtle_tf2_broadcaster',
        name='broadcaster1',
        parameters=[
            {'turtlename': 'turtle1'}
        ]
    ),

2.2 Update setup.py
~~~~~~~~~~~~~~~~~~~

Reopen ``setup.py`` and add the line so that the launch files from the ``launch/`` folder would be installed.
The ``data_files`` field should now look like this:

.. code-block:: python

    data_files=[
        ...
        (os.path.join('share', package_name, 'launch'), glob(os.path.join('launch', '*.launch.py'))),
    ],

3 Build and run
^^^^^^^^^^^^^^^

> Run ``rosdep`` in the root of your workspace to check for missing dependencies, build your updated package, and source the setup files.

Reviewer: can you add the cmd ?

Author: I specifically omitted the commands (like `rosdep install` and `colcon build`) in later tutorials as all the required steps were already mentioned in the static broadcaster tutorial. This way tutorials will cover only the necessary information and does not repeat much. Do think I should add that?

Reviewer 2: +1 for adding the command here.

I always have to look up how to run `rosdep`. I think it's fine to repeat small, convenient things like this. We're just trying to avoid writing the "how to setup a package" section, or other similar common operations, over and over. One line should be fine.

Now run the launch file that will start the turtlesim simulation node and ``turtle_tf2_broadcaster`` node:

.. code-block:: console

    ros2 launch learning_tf2_py turtle_tf2_demo.launch.py

In the second terminal window type the following command:

.. code-block:: console

    ros2 run turtlesim turtle_teleop_key

> You will now see that the turtlesim simulation have started with one turtle in the middle.

Reviewer: picture? or even better a gif ?

.. code-block:: python

    self .subscription = self.create_subscription(
        Pose,
        f'/{self.turtlename}/pose',
        self.handle_turtle_pose,
        1)

Now, we create a Transform object and give it the appropriate metadata.

#. We need to give the transform being published a timestamp, we'll just stamp it with the current time, ``ros::Time::now()``.

#. Then, we need to set the name of the parent frame of the link we're creating, in this case ``world``

#. Finally, we need to set the name of the child node of the link we're creating, in this case this is the name of the turtle itself.

The handler function for the turtle pose message broadcasts this turtle's translation and rotation, and publishes it as a transform from frame ``world`` to frame ``turtleX``.

.. code-block:: python

    t.header.stamp = self.get_clock().now().to_msg()
    t.header.frame_id = 'world'
    t.child_frame_id = self.turtlename

Here we copy the information from the 3D turtle pose into the 3D transform.

.. code-block:: python

    t.transform.translation.x = msg.x
    t.transform.translation.y = msg.y
    t.transform.translation.z = 0.0

>     q = tf_transformations.quaternion_from_euler(0, 0, msg.theta)

Author: I was actually going to cover the quaternion fundamentals in a standalone tutorial. I can add links to it once it is done. I guess I will also include the video you mentioned, it is really good.

    t.transform.rotation.x = q[0]
    t.transform.rotation.y = q[1]
    t.transform.rotation.z = q[2]
    t.transform.rotation.w = q[3]

This is where the real work is done. Sending a transform with a ``TransformBroadcaster`` requires passing in just the transform itself.

.. code-block:: python

    br.sendTransform(t)

.. note::

    You can also publish static transforms on the same pattern by instantiating a ``tf2_ros.StaticTransformBroadcaster`` instead of a ``tf2_ros.TransformBroadcaster``.

>     The static transforms will be published on the ``/tf_static`` topic and will be sent only when required (latched topic) and not periodically.

Author: I just found out that this statement is incorrect in ROS2. According to my [findings](https://docs.ros.org/en/rolling/Tutorials/Intra-Process-Communication.html?highlight=latched#latching) latched topics are not yet implemented. I will change that confusing wording to the following
`    The static transforms will be published on the ``/tf_static`` topic and will be sent only once when started and not periodically.
`

>     The static transforms will be published on the ``/tf_static`` topic and will be sent only when required (latched topic) and not periodically.

Reviewer 3: There's a long open bug about latching here: https://github.com/ros2/ros2/issues/464 .

In short, the concept of `transient_local` in ROS 2 is similar to, but not exactly the same as, `latched` topics in ROS 1.  But it is important that we keep the part about being sent when required, so I would reword the above to be:

```
The static transforms will be published on the ``/tf_static`` topic and will be sent only when required, and not periodically.
```

>     The static transforms will be published on the ``/tf_static`` topic and will be sent only when required (latched topic) and not periodically.

Author: Well noted! Thanks Chris

    For more details see :ref:`here <WritingATf2StaticBroadcasterPy>`.

1.2 Add dependencies
~~~~~~~~~~~~~~~~~~~~

Navigate one level back to the ``src/learning_tf2_py`` directory, where the ``setup.py``, ``setup.cfg``, and ``package.xml`` files are located.

Open ``package.xml`` with your text editor.
Add the following dependencies corresponding to your node's import statements:

.. code-block:: xml

    <exec_depend>launch</exec_depend>
    <exec_depend>launch_ros</exec_depend>

This declares the additional required ``launch`` and ``launch_ros`` dependencies when its code is executed.

Make sure to save the file.

1.3 Add an entry point
~~~~~~~~~~~~~~~~~~~~~~

To allow the ``ros2 run`` command to run your node, you must add the entry point
to ``setup.py`` (located in the ``src/learning_tf2_py`` directory).

Add the following line between the ``'console_scripts':`` brackets:

.. code-block:: python

>     'turtle_tf2_broadcaster = learning_tf2_py.turtle_tf2_broadcaster:main',

Reviewer 2: It would probably be nice to add the whole `console_scripts` block here, for more context.

2 Write the launch file
^^^^^^^^^^^^^^^^^^^^^^^

Now create a launch file for this demo. With your text editor, create a new
file called ``turtle_tf2_demo.launch.py``, and add the following lines:

.. code-block:: python

    from launch import LaunchDescription
    from launch_ros.actions import Node

    def generate_launch_description():
        return LaunchDescription([
            Node(
                package='turtlesim',
                executable='turtlesim_node',
                name='sim'
            ),
            Node(
                package='learning_tf2_py',
                executable='turtle_tf2_broadcaster',
                name='broadcaster1',
                parameters=[
                    {'turtlename': 'turtle1'}
                ]
            ),
        ])

2.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

> First we import required modules from the ``launch`` and ``launch_ros`` packages.

Reviewer 2: It'd be nice to mention that `launch` is a generic launching framework (not ROS specific) and `launch_ros` has ROS specific things, such as nodes.

.. code-block:: python

    from launch import LaunchDescription
    from launch_ros.actions import Node

Now we run our nodes that start the turtlesim simulation and broadcast ``turtle1`` state to the tf2 using our ``turtle_tf2_broadcaster`` node.

.. code-block:: python

    Node(
        package='turtlesim',
        executable='turtlesim_node',
        name='sim'
    ),
    Node(
        package='learning_tf2_py',
        executable='turtle_tf2_broadcaster',
        name='broadcaster1',
        parameters=[
            {'turtlename': 'turtle1'}
        ]
    ),

2.2 Update setup.py
~~~~~~~~~~~~~~~~~~~

Reopen ``setup.py`` and add the line so that the launch files from the ``launch/`` folder would be installed.
The ``data_files`` field should now look like this:

.. code-block:: python

    data_files=[
        ...
        (os.path.join('share', package_name, 'launch'), glob(os.path.join('launch', '*.launch.py'))),
    ],

> 

Reviewer 2: It'd be nice to link to some launch file tutorials here.

3 Build and run
^^^^^^^^^^^^^^^

Run ``rosdep`` in the root of your workspace to check for missing dependencies, build your updated package, and source the setup files.

Now run the launch file that will start the turtlesim simulation node and ``turtle_tf2_broadcaster`` node:

.. code-block:: console

    ros2 launch learning_tf2_py turtle_tf2_demo.launch.py

In the second terminal window type the following command:

.. code-block:: console

    ros2 run turtlesim turtle_teleop_key

> You will now see that the turtlesim simulation have started with one turtle in the middle that you can control.

Reviewer 2: ```suggestion
You will now see that the turtlesim simulation have started with one turtle that you can control.
```

.. image:: turtlesim_broadcast.png

Now, use the ``tf2_echo`` tool to check if the turtle pose is actually getting broadcast to tf2:

.. code-block:: console

    ros2 run tf2_ros tf2_echo world turtle1

This should show you the pose of the first turtle.

> Drive around the turtle using the arrow keys (make sure your terminal window is active, not your simulator window).

Reviewer 2: ```suggestion
Drive around the turtle using the arrow keys (make sure your ``turtle_teleop_key`` terminal window is active, not your simulator window).
```

.. _WritingATf2BroadcasterPy:

Writing a tf2 broadcaster (Python)
==================================

**Goal:** Learn how to broadcast the state of a robot to tf2.

**Tutorial level:** Intermediate

**Time:** 15 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

In the next two tutorials we will write the code to reproduce the demo from the :ref:`Introduction to tf2 <IntroToTf2>` tutorial.
After that, the following tutorials focus on extending the demo with more advanced tf2 features.

Prerequisites
-------------

> This tutorial assumes you have a working knowledge of working with ROS2 and you have completed the :ref:`Introduction to tf2 tutorial <IntroToTf2>`.

Reviewer 3: ```suggestion
This tutorial assumes you have a working knowledge of ROS 2 and you have completed the :ref:`Introduction to tf2 tutorial <IntroToTf2>`.
```

to ``setup.py`` (located in the ``src/learning_tf2_py`` directory).

Add the following line between the ``'console_scripts':`` brackets:

.. code-block:: python

    'console_scripts': [
        'turtle_tf2_broadcaster = learning_tf2_py.turtle_tf2_broadcaster:main',
    ],

2 Write the launch file
^^^^^^^^^^^^^^^^^^^^^^^

Now create a launch file for this demo. With your text editor, create a new
file called ``turtle_tf2_demo.launch.py``, and add the following lines:

.. code-block:: python

    from launch import LaunchDescription
    from launch_ros.actions import Node

    def generate_launch_description():
        return LaunchDescription([
            Node(
                package='turtlesim',
                executable='turtlesim_node',
                name='sim'
            ),
            Node(
                package='learning_tf2_py',
                executable='turtle_tf2_broadcaster',
                name='broadcaster1',
                parameters=[
                    {'turtlename': 'turtle1'}
                ]
            ),
        ])

2.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

First we import required modules from the ``launch`` and ``launch_ros`` packages.

> It should be noted that ``launch`` is a generic launching framework (not ROS2 specific) and ``launch_ros`` has ROS2 specific things, like nodes that we import here.

Reviewer 3: ```suggestion
It should be noted that ``launch`` is a generic launching framework (not ROS 2 specific) and ``launch_ros`` has ROS 2 specific things, like nodes that we import here.
```

        rclpy.init()
        node = FramePublisher()
        try:
            rclpy.spin(node)
        except KeyboardInterrupt:
            pass

        rclpy.shutdown()

1.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

Now, let's take a look at the code that is relevant to publishing the turtle pose to tf2.
Firstly, we define and acquire a single parameter ``turtlename``, which specifies a turtle name, e.g. ``turtle1`` or ``turtle2``.

.. code-block:: python

    self.declare_parameter('turtlename', 'turtle')
    self.turtlename = self.get_parameter(
        'turtlename').get_parameter_value().string_value

Afterward, the node subscribes to topic ``turtleX/pose`` and runs function ``handle_turtle_pose`` on every incoming message.

.. code-block:: python

    self .subscription = self.create_subscription(
        Pose,
        f'/{self.turtlename}/pose',
        self.handle_turtle_pose,
        1)

Now, we create a Transform object and give it the appropriate metadata.

#. We need to give the transform being published a timestamp, and we'll just stamp it with the current time by calling ``self.get_clock().now()``. This will return the current time used by the ``Node``.

> #. Then we need to set the name of the parent frame of the link we're creating, in this case ``world``

Reviewer: ```suggestion
#. Then we need to set the name of the parent frame of the link we're creating, in this case ``world``.
```

    from launch_ros.actions import Node

Now we run our nodes that start the turtlesim simulation and broadcast ``turtle1`` state to the tf2 using our ``turtle_tf2_broadcaster`` node.

.. code-block:: python

    Node(
        package='turtlesim',
        executable='turtlesim_node',
        name='sim'
    ),
    Node(
        package='learning_tf2_py',
        executable='turtle_tf2_broadcaster',
        name='broadcaster1',
        parameters=[
            {'turtlename': 'turtle1'}
        ]
    ),

2.2 Update setup.py
~~~~~~~~~~~~~~~~~~~

Reopen ``setup.py`` and add the line so that the launch files from the ``launch/`` folder would be installed.
The ``data_files`` field should now look like this:

.. code-block:: python

    data_files=[
        ...
        (os.path.join('share', package_name, 'launch'), glob(os.path.join('launch', '*.launch.py'))),
    ],

You can learn more about creating launch files in :ref:`this tutorial <ROS2Launch>`.

3 Build and run
^^^^^^^^^^^^^^^

Run ``rosdep`` in the root of your workspace to check for missing dependencies, build your updated package, and source the setup files.

.. tabs::

   .. group-tab:: Linux

>       .. code-block:: console

Reviewer: I can see here `.. code-block:: console` but not for the other platforms. is this right ?

Author: I believe it is correct. This part was taken from https://github.com/ros2/ros2_documentation/blob/rolling/source/Tutorials/Writing-A-Simple-Py-Publisher-And-Subscriber.rst#4-build-and-run

Author: So as they are simply a text blocks with statements `rosdep only runs on Linux, so you can skip ahead to next step.` it does not require the `.. code-block:: console` tag. I will mark it as resolved

        rosdep install -i --from-path src --rosdistro rolling -y

   .. group-tab:: macOS

>       rosdep only runs on Linux, so you can skip ahead to next step.

Reviewer: can we at least mention the required dependencies for macOS and Windows ? Otherwise it will fail when we will execute the code, right ?

Author: Yeah, you are right! These packages are listed in the `package.xml` file:
```
geometry_msgs
launch
launch_ros
rclpy
tf_transformations
tf2_ros
turtlesim
```

Should I just mention that user will need to install only these three as the others should be in the base ROS2 installation already?
```
geometry-msgs
tf-transformations
turtlesim
```

Reviewer: yes, I think you should provided there at least a list of packages. You can also mention to include them in the workspace

   .. group-tab:: Windows

      rosdep only runs on Linux, so you can skip ahead to next step.

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

      colcon build --packages-select learning_tf2_py

  .. group-tab:: macOS

    .. code-block:: console

      colcon build --packages-select learning_tf2_py

  .. group-tab:: Windows

    .. code-block:: console

      colcon build --merge-install --packages-select learning_tf2_py

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

      . install/setup.bash

  .. group-tab:: macOS

    .. code-block:: console

      . install/setup.bash

  .. group-tab:: Windows

    .. code-block:: console

>       call install/setup.bat

Reviewer: ```suggestion
      # cmd
      call install\setup.bat
      # Powershell
      .\install\setup.ps1
```
